# new Promise((resolve, reject))
```java
let promise = new Promise(function(resolve, reject) {
  // executor 
});
```
new Promise에 전달되는 함수는 executor(실행자, 실행 함수) 라고 부른다.
executor는 enw Promise가 만들어질때 자동으로 실행된다.
* resolve(value) - 성공시 실행
* reject(error) - 에러 발생시 실행

한편, new Promise 생성자가 반환하는 promise 객체는 다음과 같은 내부 프로퍼티를 갖는다.
1. state - 처음엔 'pending' (보류) 였다가 resolve가 호출되면 'fulfilled', reject가 호출되면 'rejected'로 반환.
2. result - 처음엔 undefined 였다가 resolve(value) 호출시 value, reject(error) 호출시 error 로 반환.

*state와 result는 내부 프로퍼티 이므로 직접 접근할수 없다.*

## 소비자 then, catch, finally

### then
```java
  promise.then(
    (resolve) => {
      // 결과 resolve를 반환함.
    },
    (reject) => {
      // 에러 reject를 반환함.
    }
  )
```

### catch
에러가 발생한 경우만 다루고 싶다면 .then(null, errorHandlingFunction) 또는 .catch(error)
```java
  promise.catch(alert) // alert(reject) 실행됨
```

### finally
결과가 어떻근 마무리가 필요하면 finally가 유용하다.
```java
new Promise((res, rej) => {
    res('ㅋㅋ')
  })
    .then(
      (resolve) => {
        console.log(resolve)
      },
      (reject) => {}
    )
    .finally(() => {
      console.log('무조건')
    }) // ㅋㅋ 무조건
```
